core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -1434,38 +1435,51 @@ public void executeSystemTask(WorkflowSystemTask systemTask, String taskId, long
             if (task.getStatus() == SCHEDULED || !systemTask.isAsyncComplete(task)) {
-                task.setPollCount(task.getPollCount() + 1);
+                task.incrementPollCount();
                 executionDAOFacade.updateTask(task);
             }
 
+            // load task data (input/output) from external storage, if necessary
             deciderService.populateTaskData(task);
 
-            // Stop polling for asyncComplete system tasks that are not in SCHEDULED state
-            if (systemTask.isAsyncComplete(task) && task.getStatus() != SCHEDULED) {
-                queueDAO.remove(QueueUtils.getQueueName(task), task.getTaskId());
-                return;
+            if (task.getStatus() == SCHEDULED) {
+                systemTask.start(workflow, task, this);
+            } else if (task.getStatus() == IN_PROGRESS) {
+                systemTask.execute(workflow, task, this);
             }
 
-            switch (task.getStatus()) {
-                case SCHEDULED:
-                    systemTask.start(workflow, task, this);
-                    break;
-
-                case IN_PROGRESS:
-                    systemTask.execute(workflow, task, this);
-                    break;
-                default:
-                    break;
+            if (task.getOutputData() != null && !task.getOutputData().isEmpty()) {
+                deciderService.externalizeTaskData(task);
             }
 
-            if (!task.getStatus().isTerminal()) {
+            // Update message in Task queue based on Task status
+            // Stop polling for asyncComplete system tasks that are not in SCHEDULED state
+            if (systemTask.isAsyncComplete(task) && task.getStatus() != SCHEDULED) {
+                queueDAO.remove(QueueUtils.getQueueName(task), task.getTaskId());
+            }
+            else if(task.getStatus().isTerminal()) {
+                task.setEndTime(System.currentTimeMillis());
+                queueDAO.remove(queueName, task.getTaskId());
+                LOGGER.debug("Task: {} removed from taskQueue: {} since the task status is {}", task, queueName,
+                        task.getStatus());
+            } else {
                 task.setCallbackAfterSeconds(callbackTime);
+                new RetryUtil<>().retryOnException(() -> {
+                    // postpone based on callbackTime
+                    queueDAO.postpone(queueName, task.getTaskId(), task.getWorkflowPriority(), callbackTime);
+                    LOGGER.debug(
+                            "Task: {} postponed in taskQueue: {} since the task status is {} with callbackAfterSeconds: {}",
+                            task, queueName, task.getStatus(), callbackTime);
+                    return null;
+                }, null, null, 2, "Postponing Task message in queue for taskId: " + task.getTaskId(), "postponeTaskMessage");
             }
 
-            updateTask(new TaskResult(task));
```

Comment:
@apanicker-nflx I think this call is not appropriate here, as it seems to do more than just updating the task. Please verify.

For this particular case, where we want to do retries for BACKEND_ERRORs, even if the `WorkflowSystemTask.start` leaves or sets the `Task` in `SCHEDULED` state, [updateTask](https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L964-L968) and [TaskResult constructor](https://github.com/Netflix/conductor/blob/main/common/src/main/java/com/netflix/conductor/common/metadata/tasks/TaskResult.java#L86-L88) change it to `IN_PROGRESS`. I removed the call and wrote only the essential and relevant parts of `updateTask`.